src/client.cc

```diff
@@ -117,39 +118,40 @@ DisplayCoord Client::dimensions() const
     return m_ui->dimensions();
 }
 
+String generate_context_info(const Context& context)
+{
+    String s = "";
+    if (context.buffer().is_modified())
+        s += "[+]";
+    if (context.client().input_handler().is_recording())
+        s += format("[recording ({})]", context.client().input_handler().recording_reg());
+    if (context.buffer().flags() & Buffer::Flags::New)
+        s += "[new file]";
+    if (context.hooks_disabled())
+        s += "[no-hooks]";
+    if (context.buffer().flags() & Buffer::Flags::Fifo)
+        s += "[fifo]";
+    return s;
```

Comment:
The String class takes care of all that, if its small enough (<= 22 bytes), it should end up embeded in the string object, else it will end up heap allocated, but the String is taking care of the ownership of that allocation, so this code is safe.

In Kakoune code, and generally in modern style C++, these problems are solved, its only the building block classes that should take care of memory allocation/freeing and provide a safe abstraction that prevent leaks.